### PR TITLE
Fix NullReferenceException in ProjectItemInstance.SetMetadataObject

### LINKED ISSUE #2264: Intermittent NRE when using CallTarget
Public copy of internal bug [451463](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/451463).

Builds have failed with this callstack ~20 times in the last 7 or so months in the internal VS build. There's no known simplification, so it's pretty hard to track down.

```
F:\dd\src\ConfigData\BuildTargets\Microsoft.DevDiv.targets(1601,5): error MSB4018: The "CallTarget" task failed unexpectedly. F:\dd\src\ConfigData\BuildTargets\Microsoft.DevDiv.targets(1601,5): error MSB4018: System.NullReferenceException: Object reference not set to an instance of an object.
F:\dd\src\ConfigData\BuildTargets\Microsoft.DevDiv.targets(1601,5): error MSB4018: at Microsoft.Build.Execution.ProjectItemInstance.TaskItem.SetMetadataObject(String name, String metadataValueEscaped, Boolean allowItemSpecModifiers) 
F:\dd\src\ConfigData\BuildTargets\Microsoft.DevDiv.targets(1601,5): error MSB4018: at Microsoft.Build.Execution.ProjectItemInstance.TaskItem.SetMetadata(String metadataName, String metadataValueEscaped) F:\dd\src\ConfigData\BuildTargets\Microsoft.DevDiv.targets(1601,5): error MSB4018: at Microsoft.Build.BackEnd.MSBuild.d__77.MoveNext() 
F:\dd\src\ConfigData\BuildTargets\Microsoft.DevDiv.targets(1601,5): error MSB4018: --- End of stack trace from previous location where exception was thrown --- 
F:\dd\src\ConfigData\BuildTargets\Microsoft.DevDiv.targets(1601,5): error MSB4018: at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) 
F:\dd\src\ConfigData\BuildTargets\Microsoft.DevDiv.targets(1601,5): error MSB4018: at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) 
F:\dd\src\ConfigData\BuildTargets\Microsoft.DevDiv.targets(1601,5): error MSB4018: at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task) 
F:\dd\src\ConfigData\BuildTargets\Microsoft.DevDiv.targets(1601,5): error MSB4018: at Microsoft.Build.BackEnd.TaskBuilder.d__26.MoveNext()
```

@jeffkl and I looked at the [code that's throwing](https://github.com/Microsoft/msbuild/blob/dd134d0a917b870b6a6e805984909a8c341a8083/src/Build/Instance/ProjectItemInstance.cs#L1680-L1686) on Monday. It guards against an NRE:

```c#
_directMetadata = _directMetadata ?? new CopyOnWritePropertyDictionary<ProjectMetadataInstance>();
ProjectMetadataInstance metadatum = new ProjectMetadataInstance(name, metadataValueEscaped, allowItemSpecModifiers /* may not be built-in metadata name */);
_directMetadata.Set(metadatum);
```

So it seems like a race condition, where something is [resetting](https://github.com/Microsoft/msbuild/blob/dd134d0a917b870b6a6e805984909a8c341a8083/src/Build/Instance/ProjectItemInstance.cs#L1306) metadata on an item concurrently with calling `SetMetadataObject`. That's surprising, since this code is fairly single-threaded. There are some `async` methods involved around the [legacy `CallTarget`](https://github.com/Microsoft/msbuild/blob/497d797007410f43e1d7501b7d8254313b07b2a7/src/Build/BackEnd/Components/RequestBuilder/TaskHost.cs#L844-L859) handling, which seems to be involved in every case.

### PR DESCRIPTION
**Pull request by jeffkl**
Sometimes when `SetMetadataObject` is executing, `RemoveMetadata` can set the backing collection to `null` which causes a `NullReferenceException`.  This change adds an atomic "remove if empty" operation to the `CopyOnWritePropertyDictionary<T>` which will clear the dictionary without resetting the reference to null.

Fixes #2264

### CODE REVIEW

File: src/Build/Instance/ProjectItemInstance.cs [hunk 1]
```diff
@@ -1296,16 +1295,8 @@ public void RemoveMetadata(string metadataName)
...
-                    _directMetadata.Remove(metadataName);
-
-                    // If the metadata was all removed, toss the dictionary
-                    if (_directMetadata.Count == 0)
-                    {
-                        _directMetadata = null;
-                    }
-                }
+                // If the metadata was all removed, toss the dictionary
+                _directMetadata?.Remove(metadataName, clearIfEmpty: true);
```
Comment (cdmihai) [inline src/Build/Instance/ProjectItemInstance.cs]: Why not set it to null anymore like the old code did?

File: src/Build/Instance/ProjectItemInstance.cs [hunk 2]
```diff
@@ -1235,7 +1235,6 @@ public string GetMetadataEscaped(string metadataName)
 
                 if (_directMetadata != null)
                 {
-                    Debug.Assert(_directMetadata.Count != 0, "We should not waste a dictionary for no metadata");
```
Comment (cdmihai) [inline src/Build/Instance/ProjectItemInstance.cs]: Why remove the asserts? They seem useful ...

File: src/Build/Instance/ProjectItemInstance.cs (hunk 1 above)
Comment (cdmihai) [reply src/Build/Instance/ProjectItemInstance.cs]: Ahh, I see to avoid the race over the metadata.

**Comment by jeffkl** [comment]
@dotnet-bot test Windows_NT Build for Full please

**Comment by AndyGerlicher** [comment]
Looks good for perf:

Test | Type | Overall | Significant? | Value
:--- | :---| :--- | :--- | :---
DotnetConsoleProject | Evaluation: Time (ms) | :white_check_mark: | yes | 27.1678 -> 27.1347 (-0.122%)
DotnetConsoleProject | Evaluation: Memory | :ok_hand: | no | 4723473 -> 4723555 (0.002%)
DotnetWebProject | Evaluation: Time (ms) | :white_check_mark: | yes | 41.3583 -> 40.6659 (-1.674%)
DotnetWebProject | Evaluation: Memory | ::ok_hand: | no | 6182080 -> 6182005 (-0.001%)
DotnetMvcProject | Evaluation: Time (ms) | :red_circle: | yes | 78.9432 -> 79.0985 (0.197%)
DotnetMvcProject | Evaluation: Memory | :ok_hand: | no | 8460725 -> 8460784 (0.001%)
Picasso | Evaluation: Time (ms) | :white_check_mark: | yes | 233.7267 -> 227.7558 (-2.555%)
Picasso | Evaluation: Memory | :ok_hand: | no | 32198955 -> 32199690 (0.002%)
SmallP2POldCsproj | Evaluation: Time (ms) | :red_circle: | yes | 36.7586 -> 36.9414 (0.497%)
SmallP2POldCsproj | Evaluation: Memory | ::ok_hand: | no | 5187470 -> 5187357 (-0.002%)
Generated_100_100_v150 | Evaluation: Time (ms) | :white_check_mark: | yes | 1395.2903 -> 1393.0043 (-0.164%)
Generated_100_100_v150 | Evaluation: Memory | ::ok_hand: | no | 174567744 -> 174567416 (0%)
LargeP2POldCsproj | Evaluation: Time (ms) | ::ok_hand: | no | 653.6549 -> 652.9236 (-0.112%)
LargeP2POldCsproj | Evaluation: Memory | :ok_hand: | no | 68739778 -> 68740345 (0.001%)
roslyn | Evaluation: Time (ms) | :white_check_mark: | yes | 5246.9899 -> 5223.9444 (-0.439%)
roslyn | Evaluation: Memory | ::ok_hand: | no | 579757569 -> 579755978 (0%)

**Pull request merged**